File: front end/preprocess.js
import fs from 'fs';
import path from 'path';

console.log("Starting JavaScript-based Movie Preprocessor...");

const moviesPath = path.resolve('..', 'tmdb_5000_movies.csv');
const creditsPath = path.resolve('..', 'tmdb_5000_credits.csv');

// Helper: Custom CSV Parser (Handles quotes, newlines, and escaped quotes inside JSON columns)
function parseCSV(filePath) {
  const text = fs.readFileSync(filePath, 'utf-8');
  const result = [];
  let row = [];
  let col = '';
  let inQuotes = false;
  
  for (let i = 0; i < text.length; i++) {
    const char = text[i];
    const next = text[i + 1];
    
    if (char === '"') {
      if (inQuotes && next === '"') {
        col += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (char === ',' && !inQuotes) {
      row.push(col.trim());
      col = '';
    } else if ((char === '\r' || char === '\n') && !inQuotes) {
      row.push(col.trim());
      if (row.length > 1 || (row.length === 1 && row[0] !== '')) {
        result.push(row);
      }
      row = [];
      col = '';
      if (char === '\r' && next === '\n') i++;
    } else {
      col += char;
    }
  }
  if (col !== '' || row.length > 0) {
    row.push(col.trim());
    result.push(row);
  }
  return result;
}

try {
  // Read and parse Movies CSV
  console.log("Reading and parsing tmdb_5000_movies.csv...");
  const moviesRaw = parseCSV(moviesPath);
  const moviesHeader = moviesRaw[0];
  const moviesRows = moviesRaw.slice(1);
  
  const mColIdx = {
    id: moviesHeader.indexOf('id'),
    title: moviesHeader.indexOf('title'),
    genres: moviesHeader.indexOf('genres'),
    overview: moviesHeader.indexOf('overview'),
    popularity: moviesHeader.indexOf('popularity'),
    release_date: moviesHeader.indexOf('release_date'),
    runtime: moviesHeader.indexOf('runtime'),
    vote_average: moviesHeader.indexOf('vote_average'),
    keywords: moviesHeader.indexOf('keywords')
  };

  // Read and parse Credits CSV
  console.log("Reading and parsing tmdb_5000_credits.csv...");
  const creditsRaw = parseCSV(creditsPath);
  const creditsHeader = creditsRaw[0];
  const creditsRows = creditsRaw.slice(1);
  
  const cColIdx = {
    movie_id: creditsHeader.indexOf('movie_id'),
    cast: creditsHeader.indexOf('cast'),
    crew: creditsHeader.indexOf('crew')
  };

  // Map Credits by Movie ID
  console.log("Mapping credits data...");
  const creditsMap = new Map();
  creditsRows.forEach(row => {
    const id = row[cColIdx.movie_id];
    const castRaw = row[cColIdx.cast];
    const crewRaw = row[cColIdx.crew];
    
    let cast = [];
    let crew = [];
    
    try {
      if (castRaw) {
        // Parse JSON safely
        const parsedCast = JSON.parse(castRaw);
        cast = parsedCast.slice(0, 3).map(c => c.name);
      }
    } catch (e) {}

    try {
      if (crewRaw) {
        const parsedCrew = JSON.parse(crewRaw);
        const directorObj = parsedCrew.find(c => c.job === 'Director');
        if (directorObj) {
          crew = [directorObj.name];
        }
      }
    } catch (e) {}

    creditsMap.set(id, { cast, director: crew[0] || '' });
  });

  // Preprocess Movies
  console.log("Preprocessing movies and building search tags...");
  const preprocessedMovies = [];

  // Simple Stopwords Set
  const stopwords = new Set([
    'a', 'about', 'above', 'after', 'again', 'against', 'all', 'am', 'an', 'and', 'any', 'are', 'as', 'at', 
    'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'cannot', 'could', 
    'did', 'do', 'does', 'doing', 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'has', 'have', 
    'having', 'he', 'her', 'here', 'hers', 'herself', 'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 
    'it', 'its', 'itself', 'me', 'more', 'most', 'my', 'myself', 'no', 'nor', 'not', 'of', 'off', 'on', 'once', 
    'only', 'or', 'other', 'ought', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 'same', 'she', 'should', 
    'so', 'some', 'such', 'than', 'that', 'the', 'their', 'theirs', 'them', 'themselves', 'then', 'there', 
    'these', 'they', 'this', 'those', 'through', 'to', 'too', 'under', 'until', 'up', 'very', 'was', 'we', 'were', 
    'what', 'when', 'where', 'which', 'while', 'who', 'whom', 'why', 'with', 'would', 'you', 'your', 'yours', 
    'yourself', 'yourselves'
  ]);

  moviesRows.forEach(row => {
    const id = row[mColIdx.id];
    const title = row[mColIdx.title];
    const genresRaw = row[mColIdx.genres];
    const keywordsRaw = row[mColIdx.keywords];
    const overview = row[mColIdx.overview] || '';
    const popularity = parseFloat(row[mColIdx.popularity]) || 0.0;
    const release_date = row[mColIdx.release_date] || 'Unknown';
    const runtime = parseInt(row[mColIdx.runtime]) || 0;
    const vote_average = parseFloat(row[mColIdx.vote_average]) || 0.0;

    let genres = [];
    try {
      if (genresRaw) {
        genres = JSON.parse(genresRaw).map(g => g.name);
      }
    } catch (e) {}

    let keywords = [];
    try {
      if (keywordsRaw) {
        keywords = JSON.parse(keywordsRaw).map(k => k.name);
      }
    } catch (e) {}

    // Get credits
    const credits = creditsMap.get(id) || { cast: [], director: '' };

    // Format display names
    const display_genres = genres;
    const display_cast = credits.cast;
    const display_director = credits.director;

    // Create tags for recommendation
    const cleanGenres = genres.map(g => g.replace(/\s+/g, '').toLowerCase());
    const cleanKeywords = keywords.map(k => k.replace(/\s+/g, '').toLowerCase());
    const cleanCast = credits.cast.map(c => c.replace(/\s+/g, '').toLowerCase());
    const cleanDirector = credits.director ? [credits.director.replace(/\s+/g, '').toLowerCase()] : [];
    
    // Overview words
    const overviewWords = overview
      .toLowerCase()
      .replace(/[^\w\s]/g, '')
      .split(/\s+/)
      .filter(w => w && !stopwords.has(w));

    // Combined tags string
    const tags = [
      ...overviewWords,
      ...cleanGenres,
      ...cleanKeywords,
      ...cleanCast,
      ...cleanDirector
    ];
    
    const tags_str = tags.join(' ');

    preprocessedMovies.push({
      id: parseInt(id),
      title: title,
      overview: overview,
      genres: display_genres,
      cast: display_cast,
      director: display_director,
      vote_average: vote_average,
      popularity: popularity,
      release_date: release_date,
      runtime: runtime,
      tags_str: tags_str
    });
  });

  // Write preprocessed movies as JSON asset
  const outDir = path.resolve('src');
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true });
  }
  const outPath = path.join(outDir, 'movies_data.json');
  fs.writeFileSync(outPath, JSON.stringify(preprocessedMovies, null, 2), 'utf-8');
  console.log(`Success! Preprocessed ${preprocessedMovies.length} movies and saved to ${outPath}`);

} catch (err) {
  console.error("Error during preprocessing:", err);
}

File: Backend/preprocess.py
import numpy as np
import pandas as pd
import ast
import pickle
import collections
import os
import sys
import json

print("==================================================")
print("Starting Movie Recommender System Preprocessing...")
print("==================================================")

# Determine base paths dynamically
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.abspath(os.path.join(SCRIPT_DIR, '..'))

def locate_file(filename):
    possible_paths = [
        os.path.join(SCRIPT_DIR, 'data', filename),
        os.path.join(SCRIPT_DIR, filename),
        os.path.join(PROJECT_ROOT, filename),
        os.path.join(PROJECT_ROOT, 'data', filename)
    ]
    for path in possible_paths:
        if os.path.exists(path):
            return path
    return None

movies_path = locate_file('tmdb_5000_movies.csv')
credits_path = locate_file('tmdb_5000_credits.csv')

if not movies_path or not credits_path:
    print(f"Error: Raw CSV datasets not found in {SCRIPT_DIR} or {PROJECT_ROOT}")
    sys.exit(1)

print(f"Loading raw datasets:\n - Movies: {movies_path}\n - Credits: {credits_path}")
movies = pd.read_csv(movies_path)
credits = pd.read_csv(credits_path)

# Merge datasets on title
print("Merging datasets on title...")
movies = movies.merge(credits, on='title')

# Helper converters
def convert_genres_keywords(obj):
    if isinstance(obj, list):
        return [i['name'] for i in obj if isinstance(i, dict) and 'name' in i]
    try:
        return [i['name'] for i in ast.literal_eval(obj)]
    except Exception:
        return []

def convert_cast(obj):
    if isinstance(obj, list):
        return [i['name'] for i in obj[:3] if isinstance(i, dict) and 'name' in i]
    try:
        L = []
        counter = 0
        for i in ast.literal_eval(obj):
            if counter != 3:
                L.append(i['name'])
                counter += 1
            else:
                break
        return L
    except Exception:
        return []

def fetch_director(obj):
    if isinstance(obj, list):
        for i in obj:
            if isinstance(i, dict) and i.get('job') == 'Director' and 'name' in i:
                return [i['name']]
        return []
    try:
        for i in ast.literal_eval(obj):
            if i['job'] == 'Director':
                return [i['name']]
        return []
    except Exception:
        return []

# Extract data for display
print("Cleaning columns for presentation...")
movies['display_genres'] = movies['genres'].apply(convert_genres_keywords)
movies['display_cast'] = movies['cast'].apply(convert_cast)
movies['display_director'] = movies['crew'].apply(fetch_director).apply(lambda x: x[0] if len(x) > 0 else "")

# Clean data for tag vectorization
movies['genres_clean'] = movies['display_genres'].apply(lambda x: [i.replace(" ", "").lower() for i in x])
movies['keywords_clean'] = movies['keywords'].apply(convert_genres_keywords).apply(lambda x: [i.replace(" ", "").lower() for i in x])
movies['cast_clean'] = movies['display_cast'].apply(lambda x: [i.replace(" ", "").lower() for i in x])
movies['director_clean'] = movies['display_director'].apply(lambda x: [x.replace(" ", "").lower()] if x else [])
movies['overview_clean'] = movies['overview'].apply(lambda x: x.split() if isinstance(x, str) else [])

# Merge tags
movies['tags'] = movies['overview_clean'] + movies['genres_clean'] + movies['keywords_clean'] + movies['cast_clean'] + movies['director_clean']
movies['tags_str'] = movies['tags'].apply(lambda x: " ".join(x).lower())

# Stemming
print("Applying Porter Stemmer to movie tags...")
try:
    from nltk.stem.porter import PorterStemmer
    ps = PorterStemmer()
    def stem_text(text):
        return " ".join([ps.stem(word) for word in text.split()])
    movies['tags_str'] = movies['tags_str'].apply(stem_text)
except Exception as e:
    print(f"NLTK stemming warning: {e}. Using raw lowercased tags.")

# Keep clean columns
movies_cleaned = movies[[
    'id', 'title', 'overview', 'display_genres', 'display_cast', 
    'display_director', 'vote_average', 'vote_count', 'popularity', 
    'release_date', 'runtime', 'tags_str'
]].copy()

movies_cleaned['overview'] = movies_cleaned['overview'].fillna('')
movies_cleaned['release_date'] = movies_cleaned['release_date'].fillna('Unknown')
movies_cleaned['runtime'] = movies_cleaned['runtime'].fillna(0).astype(int)

# Library-based Cosine Similarity (Scikit-Learn)
print("Computing Library Cosine Similarity matrix...")
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

cv = CountVectorizer(max_features=5000, stop_words='english')
vectors = cv.fit_transform(movies_cleaned['tags_str']).toarray()
# Optimize memory footprint by converting float64 matrix to float32
similarity_matrix = cosine_similarity(vectors).astype(np.float32)

# Custom Scratch Count Vectorizer
print("Computing Custom Scratch Vectorizer matrix...")
class CustomCountVectorizer:
    def __init__(self, max_features=5000, stop_words='english'):
        self.max_features = max_features
        self.stop_words = stop_words
        self.vocabulary_ = {}
        
    def fit_transform(self, raw_documents):
        word_counts = collections.Counter()
        stopwords = {
            'a', 'about', 'above', 'after', 'again', 'against', 'all', 'am', 'an', 'and', 'any', 'are', 'as', 'at', 
            'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'cannot', 'could', 
            'did', 'do', 'does', 'doing', 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'has', 'have', 
            'having', 'he', 'her', 'here', 'hers', 'herself', 'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 
            'it', 'its', 'itself', 'me', 'more', 'most', 'my', 'myself', 'no', 'nor', 'not', 'of', 'off', 'on', 'once', 
            'only', 'or', 'other', 'ought', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 'same', 'she', 'should', 
            'so', 'some', 'such', 'than', 'that', 'the', 'their', 'theirs', 'them', 'themselves', 'then', 'there', 
            'these', 'they', 'this', 'those', 'through', 'to', 'too', 'under', 'until', 'up', 'very', 'was', 'we', 'were', 
            'what', 'when', 'where', 'which', 'while', 'who', 'whom', 'why', 'with', 'would', 'you', 'your', 'yours', 
            'yourself', 'yourselves'
        }
        
        for doc in raw_documents:
            tokens = doc.lower().split()
            filtered_tokens = [t for t in tokens if t.isalnum() and (self.stop_words != 'english' or t not in stopwords)]
            word_counts.update(filtered_tokens)
            
        top_words = [word for word, count in word_counts.most_common(self.max_features)]
        self.vocabulary_ = {word: idx for idx, word in enumerate(top_words)}
        
        vectors = []
        for doc in raw_documents:
            tokens = doc.lower().split()
            doc_vec = [0] * len(self.vocabulary_)
            for t in tokens:
                if t in self.vocabulary_:
                    doc_vec[self.vocabulary_[t]] += 1
            vectors.append(doc_vec)
            
        return np.array(vectors, dtype=np.uint16)

ccv = CustomCountVectorizer(max_features=5000)
custom_vectors = ccv.fit_transform(movies_cleaned['tags_str'])

# Prepare output directories
backend_models_dir = os.path.join(SCRIPT_DIR, 'models')
backend_data_dir = os.path.join(SCRIPT_DIR, 'data')
os.makedirs(backend_models_dir, exist_ok=True)
os.makedirs(backend_data_dir, exist_ok=True)

# Save Pickles in Backend/models/ as well as Root
save_locations = [
    backend_models_dir,
    PROJECT_ROOT
]

for loc in set(save_locations):
    print(f"Saving pickled models to: {loc}")
    with open(os.path.join(loc, 'movies_cleaned.pkl'), 'wb') as f:
        pickle.dump(movies_cleaned, f)
    with open(os.path.join(loc, 'similarity_matrix.pkl'), 'wb') as f:
        pickle.dump(similarity_matrix, f)
    with open(os.path.join(loc, 'custom_vectors.pkl'), 'wb') as f:
        pickle.dump(custom_vectors, f)

# Export JSON for Frontend sync
json_records = []
for idx, row in movies_cleaned.iterrows():
    json_records.append({
        "id": int(row['id']),
        "title": str(row['title']),
        "overview": str(row['overview']),
        "genres": list(row['display_genres']),
        "cast": list(row['display_cast']),
        "director": str(row['display_director']),
        "vote_average": float(row['vote_average']),
        "popularity": float(row['popularity']),
        "release_date": str(row['release_date']),
        "runtime": int(row['runtime']),
        "tags_str": str(row['tags_str'])
    })

frontend_src_json = os.path.join(PROJECT_ROOT, 'front end', 'src', 'movies_data.json')
backend_data_json = os.path.join(backend_data_dir, 'movies_data.json')

with open(backend_data_json, 'w', encoding='utf-8') as f:
    json.dump(json_records, f, indent=2)

if os.path.exists(os.path.dirname(frontend_src_json)):
    with open(frontend_src_json, 'w', encoding='utf-8') as f:
        json.dump(json_records, f)
    print(f"Exported JSON dataset to frontend: {frontend_src_json}")

print("==================================================")
print("Preprocessing completed successfully!")
print(f"Total Movies Processed: {len(movies_cleaned)}")
print("==================================================")
